packages/commonwealth/client/scripts/views/pages/LaunchToken/steps/TokenInformationStep/TokenInformationForm/TokenInformationForm.tsx

```diff
@@ -25,25 +39,91 @@ import { tokenInformationFormValidationSchema } from './validation';
 const TokenInformationForm = ({
   onSubmit,
   onCancel,
+  onAddressSelected,
+  selectedAddress,
 }: TokenInformationFormProps) => {
+  const user = useUserStore();
+  const [baseOption] = communityTypeOptions;
+
+  const shouldSubmitOnAddressSelection = useRef(false);
+  const formMethodsRef = useRef<CWFormRef>();
+  const [
+    isNewTokenCommunityAdminModalOpen,
+    setIsNewTokenCommunityAdminModalOpen,
+  ] = useState(false);
+  const [isAuthModalOpen, setIsAuthModalOpen] = useState(false);
   const [isProcessingProfileImage, setIsProcessingProfileImage] =
     useState(false);
 
   const { isAddedToHomeScreen } = useAppStatus();
 
-  const [baseOption] = communityTypeOptions;
-
   const { trackAnalytics } = useBrowserAnalyticsTrack<
     MixpanelLoginPayload | BaseMixpanelPayload
   >({
     onAction: true,
   });
 
-  const handleSubmit = (values: FormSubmitValues) => {
-    // TODO 8705: call token launch endpoint
-    console.log('values => ', values);
-    onSubmit();
-  };
+  const { mutateAsync: launchToken } = useLaunchTokenMutation();
+
+  const openAddressSelectionModal = useCallback(() => {
+    if (selectedAddress) {
+      setIsAuthModalOpen(false);
+      setIsNewTokenCommunityAdminModalOpen(false);
+    } else {
+      setIsAuthModalOpen(!user.isLoggedIn);
+      setIsNewTokenCommunityAdminModalOpen(user.isLoggedIn);
+    }
+  }, [selectedAddress, user.isLoggedIn]);
+
+  useRunOnceOnCondition({
+    callback: openAddressSelectionModal,
+    shouldRun: true,
+  });
+
+  const handleSubmit = useCallback(
+    async (values: FormSubmitValues) => {
+      // get address from user
+      if (!selectedAddress) {
+        openAddressSelectionModal();
+        shouldSubmitOnAddressSelection.current = true;
+        return;
+      }
+
+      // get base chain node info
+      const nodes = fetchCachedNodes();
+      const baseNode = nodes?.find(
+        (n) => n.ethChainId === commonProtocol.ValidChains.Base,
+      );
+      if (!baseNode || !baseNode.ethChainId) {
+        notifyError('Could not find base chain node');
+        return;
+      }
+
+      // call endpoint
+      const payload = {
+        chainRpc: baseNode.url,
+        ethChainId: baseNode.ethChainId,
+        name: values.tokenName.trim(),
+        symbol: values.tokenTicker.trim(),
+        walletAddress: selectedAddress.address,
+        // TODO 9207: where to store values.tokenDescription and values.tokenImageURL
+      };
+      await launchToken(payload).catch(console.error);
+
+      onSubmit(values);
+    },
+    [openAddressSelectionModal, selectedAddress, onSubmit, launchToken],
+  );
+
+  useEffect(() => {
+    if (shouldSubmitOnAddressSelection.current) {
```

Comment:
can you explain whats going on here? + maybe add some comment as it looks too complicaed

Reply:
added a comment